### PR TITLE
KVS implementation for darwin

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem
Implement KVS for Darwin platforms

 #### Summary of Changes
 Core-data implementation for darwin.
 
 Would really use some apple-developer feedback, at least in these areas:
  
  - I programatically created every model since I was unsure if the xcode-UI generated data could be used. Painful to type
  - Unsure about object lifetime in objective-c. Do they auto-release or am I missing a ton of release calls in this PR (am I leaking memory? how to check)
  
  I have not programmed objective C in a very very long time and it shows.

### CODE REVIEW

File: src/platform/Darwin/KeyValueStoreManagerImpl.mm [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+
+        KeyValueStoreManagerImpl KeyValueStoreManagerImpl::sInstance;
+
+        CHIP_ERROR KeyValueStoreManagerImpl::Init(const char * fileName)
+        {
+            ReturnErrorCodeIf(gContext != nullptr, CHIP_ERROR_INCORRECT_STATE);
+            ReturnErrorCodeIf(fileName == nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+
+            // make sure the path is relative
+            while (fileName[0] == '/') {
```
[the remark this answers is not in the document]
Comment (andy31415) [reply src/platform/Darwin/KeyValueStoreManagerImpl.mm]: Updated - relative paths are from documents (I figured this is better than using 'current directory' and finding special folders is platform specific), but absolute paths are preserved.

**Review by msandstedt — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

File: src/platform/Darwin/KeyValueStoreManagerImpl.mm [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+                [entity setManagedObjectClassName:@"KeyValueItem"];
+
+                // create the attributes
+                NSMutableArray * properties = [NSMutableArray array];
+
+                NSAttributeDescription * keyAttribute = [[NSAttributeDescription alloc] init];
+                [keyAttribute setName:@"key"];
+                [keyAttribute setAttributeType:NSStringAttributeType];
+                [keyAttribute setOptional:NO];
+                // [keyAttribute setIndexed:YES];
```
Comment (shana-apple) [inline src/platform/Darwin/KeyValueStoreManagerImpl.mm:81]: removed commented out code

File: src/platform/Darwin/KeyValueStoreManagerImpl.mm [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+                [properties addObject:keyAttribute];
+
+                NSAttributeDescription * valueAttribute = [[NSAttributeDescription alloc] init];
+                [valueAttribute setName:@"value"];
+                [valueAttribute setAttributeType:NSBinaryDataAttributeType];
+                [valueAttribute setOptional:NO];
+                // [valueAttribute setIndexed:NO];
```
Comment (shana-apple) [inline src/platform/Darwin/KeyValueStoreManagerImpl.mm:88]: remove commented out code

**Review by woody-apple — APPROVED** [review]

**Comment by woody-apple** [comment]
@mspang @saurabhst ?

**Review by mspang — APPROVED** [review]
rubber stamp

**Pull request merged**